code/game/objects/items/RCD.dm

```diff
@@ -1157,56 +1182,152 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
 )
 
 ///pretty much rcd_create, but named differently to make myself feel less bad for copypasting from a sibling-type
-/obj/item/construction/plumbing/proc/create_machine(atom/A, mob/user)
-	if(!machinery_data || !isopenturf(A))
+/obj/item/construction/plumbing/proc/create_machine(atom/destination, mob/user)
+	if(!machinery_data || !isopenturf(destination))
 		return FALSE
 
+	if(!canPlace(destination))
+		var/obj/blueprint_type = blueprint
+		to_chat(user, span_notice("There is something blocking you from placing a [initial(blueprint_type.name)] there."))
+		return
 	if(checkResource(machinery_data["cost"][blueprint], user) && blueprint)
 		//"cost" is relative to delay at a rate of 10 matter/second  (1matter/decisecond) rather than playing with 2 different variables since everyone set it to this rate anyways.
-		if(do_after(user, machinery_data["cost"][blueprint], target = A))
-			if(checkResource(machinery_data["cost"][blueprint], user) && canPlace(A))
+		if(do_after(user, machinery_data["cost"][blueprint], target = destination))
+			if(checkResource(machinery_data["cost"][blueprint], user) && canPlace(destination))
 				useResource(machinery_data["cost"][blueprint], user)
 				activate()
-				playsound(src.loc, 'sound/machines/click.ogg', 50, TRUE)
-				new blueprint (A, FALSE, layers[current_layer])
+				playsound(loc, 'sound/machines/click.ogg', 50, TRUE)
+				if(ispath(blueprint, /obj/machinery/duct))
+					var/is_omni = current_color == DUCT_COLOR_OMNI
+					new blueprint(destination, FALSE, GLOB.pipe_paint_colors[current_color], GLOB.plumbing_layers[current_layer], null, is_omni)
+				else
+					new blueprint(destination, FALSE, GLOB.plumbing_layers[current_layer])
 				return TRUE
 
-/obj/item/construction/plumbing/proc/canPlace(turf/T)
-	if(!isopenturf(T))
+/obj/item/construction/plumbing/proc/canPlace(turf/destination)
+	if(!isopenturf(destination))
 		return FALSE
 	. = TRUE
-	for(var/obj/O in T.contents)
-		if(O.density) //let's not built ontop of dense stuff, like big machines and other obstacles, it kills my immershion
+
+	var/obj/blueprint_template = blueprint
+	var/layer_id = GLOB.plumbing_layers[current_layer]
+
+	for(var/obj/machinery/machine in destination.contents)
+		// Let's not built ontop of dense stuff, if this is also dense.
```

Comment:
is there a reason we went from checking to see if there's anything dense in the tile to just if there's a dense machine in the tile?

Reply:
I'm sure I had a reason in my head, but I don't remember what it was now (maybe I was thinking of making the check more complex?). I'll change it back to /obj real quick since I can't justify it.

Reply:
Changed to /obj; none of the code relied on it being machines.